asapdiscovery-data/asapdiscovery/data/state_expanders/state_expander.py

```diff
@@ -0,0 +1,164 @@
+import abc
+from typing import Literal
+
+# import networkx as nx
+from asapdiscovery.data.schema_v2.ligand import Ligand
+from pydantic import BaseModel, Field
+
+
+class StateExpanderBase(abc.ABC, BaseModel):
+    expander_type: Literal["StateExpanderBase"] = Field(
+        "StateExpanderBase", description="The type of expander."
+    )
+
+    @abc.abstractmethod
+    def _expand(self, ligands: list[Ligand], unique: bool = False) -> list[Ligand]:
+        ...
+
+    def expand(self, ligands: list[Ligand], unique: bool = True) -> list[Ligand]:
+        expanded_ligands = self._expand(ligands=ligands)
+        if unique:
+            return list(set(expanded_ligands))
+        else:
+            return expanded_ligands
+
+    @abc.abstractmethod
+    def _provenance(self) -> dict[str, str]:
+        """Return the software used to perform the state expansion in the workflow."""
+        ...
+
+    def provenance(self) -> dict[str, str]:
+        """
+        Get the provenance of the software and settings used to expand the molecule state.
+        Returns
+        -------
+            A dict of the expander and the software used to do the expansion.
+        """
+        data = {"expander": self.dict()}
+        data.update(self._provenance())
+        return data
+
+
+class StateExpansion(BaseModel):
+    parent: Ligand = Field(..., description="The parent ligand")
+    children: list[Ligand] = Field(
+        ..., description="The children ligands resulting from expansion"
+    )
+    expansion: Literal["stereo", "charge"] = Field(
+        ...,
+        description="The type of state expansion, this will be used "
+        "to group the expansions.",
+    )
+
+    class Config:
+        allow_mutation = False
+
+    @property
+    def n_expanded_states(self) -> int:
+        return len(self.children)
+
+
+#
+#     def to_networkx(self) -> nx.DiGraph:
+#         graph = nx.DiGraph()
+#         graph.add_node(self.parent)
+#         for child in self.children:
+#             graph.add_node(child)
+#             graph.add_edge(self.parent, child)
+#         return graph
+
+
+class StateExpansionSet(BaseModel):
```

Comment:
If you want we can remove this and just return two lists, for assigned expansions and unassigned, but up to you if you think this has value (probably does?).

Reply:
I think it makes sense to have a way to access this information, I would just maybe add a docstring to the object since it's not that easy to follow what it is doing and how it can be used.